torch/cuda/amp/amp_scaler.py

```diff
@@ -0,0 +1,332 @@
+import torch
+from collections import defaultdict
+from torch._six import container_abcs
+
+
+class _PerDeviceHelper(object):
+    """
+    Helper class to serve copies of a master tensor that may be needed on multiple devices.
+    Intended to be used as a temporary.
+    """
+    def __init__(self, master_tensor):
+        self.master = master_tensor
+        self._per_device_tensors = {}
+
+    def get(self, device):
+        retval = self._per_device_tensors.get(device, None)
+        if retval is None:
+            retval = self.master.clone() if (device == self.master.device) else self.master.to(device)
+            self._per_device_tensors[device] = retval
+        return retval
+
+
+class AmpScaler(object):
+    """
+    ``torch.float16`` has a limited dynamic range.  Late in training, when gradient magnitudes become small,
```

Comment:
When training a network in mixed precision (see <doc link>), gradients may become too small to represent as torch.float16, AKA 'half,' values. These gradients would underflow, but scaling gradients during backwards prevents this. The gradients are then unscaled before accumulation. 

Something like that?